### PR TITLE
Added Comparison tests for Faster Fifo Queue vs Multiprocessing Queue

### PR DESCRIPTION
**Pull request by tushartk**

### CODE REVIEW

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+def produce_msgs(q, p_idx, num_messages):
+    i = 0
+    while i < num_messages:
+        try:
+            q.put(make_msg(i), timeout=0.01)
+            if i % 100000 == 0:
+                log.info('Produce: %d %d', i, p_idx)
+            i += 1
+        except Full:
+            # time.sleep(0.001)
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:33]: remove this sleep comment

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            pass
+        except Exception as exc:
+            log.exception(exc)
+    log.info('Done producing! %d', p_idx)
+
+
+def consume_msgs(q, queue_type, p_idx, consume_many=None, total_num_messages=int(1e9)):
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:40]: I don't think total_num_messages has any meaning in this function. You should remove it I guess. You don't even pass any value for this parameter

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    messages_received = 0
+    while True:
+        try:
+            msgs = q.get_many(timeout=0.01, max_messages_to_get=consume_many) if queue_type == 'ff' else q.get()
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:44]: I think you should write this differently. First of all, why use a string for the queue type? Just use the class itself.
Name it queue_cls, and then you can directly call it like queue_cls() to create a queue object. This is more elegant.

Second, one of the points of these tests is to show that our queue and multiprocessing.Queue can be used interchangeably. AND we have an extra option to get multiple messages with get_many(). So I would reorganize it like this:

if consume_many == 1:
    msg = q.get()
    msgs = [msg]
else:
    msgs = q.get_many(max_messages_to_get=consume_many)

This way you don't need to check for queue type also later when you calculate messages_received
Also, use 1 as default value for consume_many instead of 1. It works perfectly well semantically.

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            if queue_type == 'mp' and msgs is None:
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:45]: I agree that sending a None message is the easiest way to terminate the consumer after all messages are 100% processed. I think we should use this for both our and mp queues.
It works perfectly for 1-message-at-a-time case, and for consume_many case you should just send num_consumers * consume_many messages to the queue, this should be enough to terminate all processes.

Alternatively you can use get() with small timeout, and then use something like multiprocessing.Event to signal that consumer processes should be terminated.

Bottom line: let's use a single termination mechanism for both cases. None msg is fine. You can write something like:
```
if any(m is None for m in msgs):
    break
```
this should work for all cases.

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+                break
+            messages_received += len(msgs) if queue_type == 'ff' else 1
+            if messages_received >= total_num_messages:
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:48]: Again, I don't see why you need messages_recieved here. Remove it along with total_num_messages

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                break
+        except Empty:
+            if q.empty():
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:51]: This is incorrect. It is ill-advised to exit like this. The fact that queue is momentarily empty does not mean that processing is finished. It may just be the case that consumers overtake producers.
Use a None mechanism or a mp.Event mechanism and remove this.
Just `pass` when `except Empty`, it is not an error.

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    for j in range(num_consumers):
+        p = multiprocessing.Process(target=consume_msgs, args=(q, queue_type, j, consume_many))
+        consumers.append(p)
+    for p in producers:
+        p.start()
+    for c in consumers:
+        c.start()
+    for p in producers:
+        p.join()
+    if queue_type == 'mp':
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:75]: Again, let's do this for all cases.

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        for j in range(num_consumers):
+            q.put(None)
+    for c in consumers:
+        c.join()
+    q.close()
+    log.info('Exiting queue type ' + queue_type)
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:81]: Use percent-formatting %s or %r in this case instead of string concatenation when you do logging

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    end_time = time()
+    return round(end_time - start_time, 5)
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:83]: Why round? I don't think you should round the number. If you want to print it later, use %-formatting.
Such as `log.debug('.5f', time)` will print time to 5 decimal places

File: cpp_faster_fifo/tests/comparison_tests.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        time_mp = process_prod_and_cons(queue_type='mp', num_producers=20, num_consumers=3, msgs_per_prod=25000,
+                                        consume_many=None)
+        self.assertLess(time_ff, time_mp)
+
+    def test_twenty_prod_to_twenty_cons(self):
+        time_ff = process_prod_and_cons(queue_type='ff', num_producers=20, num_consumers=20, msgs_per_prod=25000,
+                                        consume_many=100)
+        time_mp = process_prod_and_cons(queue_type='mp', num_producers=20, num_consumers=20, msgs_per_prod=25000,
+                                        consume_many=None)
+        self.assertLess(time_ff, time_mp)
```
Comment (alex-petrenko) [inline cpp_faster_fifo/tests/comparison_tests.py:173]: I would format it differently.
Instead of having 2 test classes let's have one class, and three function calls in each method
Two for FF one for MP.
The asserts be like this:
```
self.assertLess(time_ff, time_mp)
self.assertLess(time_ff_many, time_mp)
```

time_ff_many ought not be less than time_ff, although usually it will probably be less. Here we're just proving the point that FF is better than MP

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 1 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:33]: Done

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 2 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:40]: Done

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 5 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:48]: Done

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 4 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:45]: Done

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 3 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:44]: Done

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 6 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:51]: Done

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 7 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:75]: Done

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 8 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:81]: Done

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 9 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:83]: Done

File: cpp_faster_fifo/tests/comparison_tests.py (hunk 10 above)
Comment (tushartk) [reply cpp_faster_fifo/tests/comparison_tests.py:173]: Done

**Comment by alex-petrenko** [comment]
Great work!
